### PR TITLE
SunPy Coordinates

### PR DESCRIPTION
**Pull request by Cadair**
# SunPy Coordinates Module

This PR is the implementation of the common solar physics physical coordinate frames into the [astropy.coordinates](http://docs.astropy.org/en/stable/coordinates/index.html) framework.

Before we get started some definitions from [APE 5](https://github.com/astropy/astropy-APEs/blob/master/APE5.rst):

* A "Coordinate Representation" is a particular way of describing a unique
  point in a vector space. (Here, this means three-dimensional space, but future
  extensions might have different dimensionality, particularly if relativistic
  effects are desired.)  Examples include Cartesian coordinates, cylindrical
  polar, or latitude/longitude spherical polar coordinates.

* A "Reference System" is a scheme for orienting points in a space and
  describing how they transforms to other systems.Examples include the ICRS,
  equatorial coordinates with mean equinox, or the WGS84 geoid for
  latitude/longitude on the Earth.

* A "Coordinate Frame", "Reference Frame", or just "Frame" is a specific
  realization of a reference system - e.g., the ICRF, or J2000 equatorial
  coordinates.For some systems, there may be only one meaningful frame, while
  others may have many different frames (differentiated by something like a
  different equinox, or a different set of reference points).

* A "Coordinate" is a combination of all of the above that specifies a unique
  point.

## What is in this PR?

1. Five coordinate frames
    * Helioprojective (Cartesian)
    * HelioprojectiveRadial
    * Heliocentric
    * HeliographicStonyhurst
    * HeliographicCarrington
1. Transformations between these frames.
1. Representation objects to wrap longitude to -180 -> 180 degrees.
1. A TimeFrameAttribute subclass to allow the parsing of all time strings using `parse_time`.
1. A mapping  between Astropy WCS objects and SunPy Coordinate frames.
1. Tests
1. Documentation

## What is **not** in this PR?

1. Independant accuracy tests, currently there are tests that compare this to `sunpy.wcs`.
1. Unit tests for the transformation functions.
1. Tests for the `HelioprojectiveRadial` class (WIP).
1. Any integration with any other part of the SunPy code base.

## What does it DO?!

Anything from [calculating the coordinates of all points in a Map](https://gist.github.com/Cadair/9b5b29aebbe3e7e10e38) to [Converting coordinates from AIA to STEREO EUVI](https://gist.github.com/Cadair/cc23d737f1124470e57d).

It also allows you to do this:

```python
import matplotlib.pyplot as plt
import sunpy.coordinates
import sunpy.map
from sunpy.data.sample import AIA_171_ROLL_IMAGE

amap = sunpy.map.Map(AIA_171_ROLL_IMAGE)
im = amap.plot()
ax = plt.gca()
overlay = ax.get_coords_overlay('heliographicstonyhurst')
overlay.grid(color='red', linestyle='solid')
```
figure_1

## How do I use it?

Check out the documentation here: http://cadair-sunpy.readthedocs.org/en/coordinates-rebased/code_ref/coordinates.html

Last, but not least, thanks to @PritishC, who wrote most of it.

### CODE REVIEW

**Comment by CyclingNinja** [comment]
Note to @Cadair time can't be an array!

**Comment by Cadair** [comment]
@sunpy/sunpy-developers Review please!!

File: doc/source/code_ref/coordinates.rst [hunk 1]
```diff
@@ -0,0 +1,196 @@
+SunPy coordinates
+=================
+
+The SunPy coordinates submodule is an implementation of the common solar physics
+coordinate frames in to the :ref:`Astropy coordinates system <astropy:astropy-coordinates>`.
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: in to the -> using the

Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: coordinates system -> coordinates framework

File: doc/source/code_ref/coordinates.rst [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+The easiest interface to the coordinates module is through the `~astropy.coordinates.SkyCoord` class::
+
+  >>> import astropy.units as u
+  >>> from astropy.coordinates import SkyCoord
+  >>> import sunpy.coordinates
+  >>> c = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame='helioprojective')
+  >>> c = SkyCoord(90*u.deg, -76*u.deg, frame='helioprojectiveradial')
+  >>> c = SkyCoord(x=-72241.0*u.km, y=361206.1*u.km, z=589951.4*u.km, frame='heliocentric')
+  >>> c = SkyCoord(70*u.deg, -30*u.deg, frame='heliographicstonyhurst')
+  >>> c
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Wondering whether an underscore would improve readability in ``heliographicstonyhurst`` -> ``heliographic_stonyhurst``

File: doc/source/code_ref/coordinates.rst [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  <SkyCoord (HelioGraphicStonyhurst: dateobs=None): (lon, lat, rad) in (deg, deg, km)
+      (70.0, -30.0, 695508.0)>
+
+
+SunPy implements supports for the following solar physics coordinate systems:
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: support

File: doc/source/code_ref/coordinates.rst [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+* Helioprojective (Cartesian) `~sunpy.coordinates.frames.HelioProjective`
+* Helioprojective (Radial) `~sunpy.coordinates.frames.HelioProjectiveRadial`
+* Heliocentric `~sunpy.coordinates.frames.HelioCentric`
+* Heliographic Stonyhurst `~sunpy.coordinates.frames.HelioGraphicStonyhurst`
+* Heliographic Carrington `~sunpy.coordinates.frames.HelioGraphicCarrington`
+
+for a complete description of these frames, see `sunpy.coordinates.frames`.
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Is it worth also pointing to reference papers/documents describing these?

File: doc/source/code_ref/coordinates.rst [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+
+`~astropy.coordinates.SkyCoord` and all other `~astropy.coordinates` objects
+also support array coordinates. These work the same as single-value coordinates,
+but they store multiple coordinates in a single object. When you're going to
+apply the same operation to many different coordinates, this is a better choice
+than a list of `~astropy.coordinates.SkyCoord` objects, because it will be
+*much* faster than applying the operation to each
+`~astropy.coordinates.SkyCoord` in a for loop.
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: in a ``for`` loop (with backticks)

File: doc/source/code_ref/coordinates.rst [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+   >>> c[0]
+   <SkyCoord (HelioProjective: D0=149597870.7 km, dateobs=None, L0=0.0 deg, B0=0.0 deg, RSun=695508.0 km): (Tx, Ty) in arcsec
+       (-500.0, 100.0)>
+
+
+Accessing Coordinates
+^^^^^^^^^^^^^^^^^^^^^
+
+Individual coordinates can be accessed via attributes on the SkyCoord object,
+the names of the components of the coordinates for each frame differ. For a full
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: add 'but', i.e. 'but the names of the components'

File: doc/source/code_ref/coordinates.rst [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+description of all the properties of the frames see `sunpy.coordinates.frames`.
+
+``HelioProjective``::
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Make these into section headings at the same time as being code?

File: doc/source/code_ref/coordinates.rst [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+  >>> c = SkyCoord(-500*u.arcsec, 100*u.arcsec, frame='helioprojective')
+  >>> c.Tx
+  <Longitude180 -500.0 arcsec>
+  >>> c.Ty
+  <Latitude 100.0 arcsec>
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Worth adding a sentence to say what ``Tx`` and ``Ty`` are? Also, why use uppercase letters in the attribute names?

File: doc/source/code_ref/coordinates.rst [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+``HeliographicStonyhurst``::
+
+   >>> c = SkyCoord(70*u.deg, -30*u.deg, frame='heliographicstonyhurst')
+   >>> c.lat
+   <Latitude -30.0 deg>
+   >>> c.lon
+   <Longitude180 70.0 deg>
+   >>> c.rad
+   <Distance 695508.0 km>
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: There are 5 supported frames listed about, but 3 here - 2 missing?

File: doc/source/code_ref/coordinates.rst [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+Design of the Coordinates Module
+--------------------------------
+
+This module works by defining a collection of ``Frames``
+(`sunpy.coordinates.frames`), which exists on a transformation graph, where the
+transformations between the coordinate frames are then defined and registered
+with the transformation graph (`sunpy.coordinates.transformations`).
+
+Positions within these ``Frames`` are stored as a ``Representation`` of a
+coordinate, a representation is a description of a point in a Cartesian,
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: a representation *being* a description

File: doc/source/code_ref/coordinates.rst [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+   <Distance 695508.0 km>
+
+
+Design of the Coordinates Module
+--------------------------------
+
+This module works by defining a collection of ``Frames``
+(`sunpy.coordinates.frames`), which exists on a transformation graph, where the
+transformations between the coordinate frames are then defined and registered
+with the transformation graph (`sunpy.coordinates.transformations`).
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Worth mentioning whether (if this is the case) the graph of the transforms here is connected to the Astropy ones?

File: doc/source/code_ref/coordinates.rst [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+provides for caching Frame attributes. Frame attributes are extra data specified
+with a frame, some examples in `sunpy.coordinates` are ``dateobs`` or ``L0`` and
+``B0`` for observer location. Only the frames where this data is meaningful have
+these attributes, i.e. only the Helioprojective frames have ``L0`` and ``B0``.
+However, when you transform into another frame and then back to a projective
+frame using `SkyCoord` it will remember the attributes previously provided, and
+repopulate the final frame with them. If you were to do transformations using
+the Frames alone this would not happen.
+
+The most important implication for this in `sunpy.coordinates` is the ``RSun``
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: PEP8-wise, ``RSun`` should be ``rsun`` :)

File: doc/source/code_ref/coordinates.rst [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+parameter in the projective frames. If you create a projective frame with a
+``RSun`` attribute, if you convert back to a projective frame it will be set
+correctly. It should also be noted that, if you create a Heliographic frame and
+then transform to a projective frame with an ``RSun`` attribute, it will not
+match the ``rad`` coordinate in the Heliographic frame. This is because you may
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: Is there a reason for not using ``radius`` instead of ``rad``? (worried it could get confused with radians)

File: doc/source/code_ref/coordinates.rst [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+mean to be describing a point above the defined 'surface' of the Sun.
+
+
+Coordinates and WCS
+-------------------
+
+The `sunpy.coordinates` package provides a mapping between FITS-WCS CTYPE
+convention and the coordinate frames as defined in `sunpy.coordinates`. This is
+provided in the `sunpy.coordinates.wcs_utils` module and adds the mappings to
+the `astropy.wcs.utils.WCS_FRAME_MAPPINGS` list. This list is used by packages
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: This is quite advanced for user docs, and is an implementation detail - maybe just say that ``wcs_to_celestial_frame`` from Astropy recognizes the frames from Sunpy when sunpy.coordinates is imported?

File: doc/source/code_ref/coordinates.rst [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    :headings: ^#
+
+.. automodapi:: sunpy.coordinates.wcs_utils
+    :headings: ^#
+
+
+Attribution
+-----------
+
+Some of this documentation was borrowed from Astropy under the terms of the `BSD
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: borrowed -> adapted

File: doc/source/code_ref/coordinates.rst [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+License
+<https://raw.githubusercontent.com/astropy/astropy/master/licenses/LICENSE.rst>`_.
+
+This package was developed by Pritish Chakraborty as part of GSOC 2014.
```
Comment (astrofrog) [inline doc/source/code_ref/coordinates.rst]: And Stuart Mumford? :wink:

File: sunpy/coordinates/__init__.py [hunk 17]
```diff
@@ -0,0 +1,6 @@
+from __future__ import absolute_import
```
Comment (astrofrog) [inline sunpy/coordinates/__init__.py]: Just out of curiosity, any reason to not include ``division``? Also, some files below are missing the future import, I think you should always include it (especially division)

File: sunpy/coordinates/frameattributes.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    See the `~astropy.coordinates.FrameAttribute` API doc for further
+    information.
+
+    Parameters
+    ----------
+    default : object
+        Default value for the attribute if not provided
+    secondary_attribute : str
+        Name of a secondary instance attribute which supplies the value
+        if ``default is None`` and no value was supplied during initialization.
```
Comment (astrofrog) [inline sunpy/coordinates/frameattributes.py]:
```
if ``default`` is `None` and no value
```

File: sunpy/coordinates/frameattributes.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            return None, False
+
+        elif value == 'now':
+            return Time(datetime.datetime.now()), True
+
+        elif isinstance(value, Time):
+            out = value
+            converted = False
+
+        elif isinstance(value, basestring):
```
Comment (astrofrog) [inline sunpy/coordinates/frameattributes.py]: Important for Python 3 compat, this should be ``six.string_types`` in future (but guess you don't support Python 3 yet?)

File: sunpy/coordinates/frames.py [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+from astropy.coordinates.representation import (CartesianRepresentation,
+                                                UnitSphericalRepresentation,
+                                                SphericalRepresentation)
+from astropy.coordinates.baseframe import (BaseCoordinateFrame,
+                                           RepresentationMapping)
+from astropy.coordinates import FrameAttribute
+
+# SunPy imports
+from sunpy import sun  # For Carrington rotation number
+from representation import (SphericalWrap180Representation,
```
Comment (astrofrog) [inline sunpy/coordinates/frames.py]:
```
from .representation
```

File: sunpy/coordinates/frames.py [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+    A coordinate or frame in the Stonyhurst Heliographic
+    system.
+
+    This frame has its origin at the solar centre and the north pole above the
+    solar north pole, and the zero line on longitude pointing towards the
+    Earth.
+
+    Parameters
+    ----------
+    representation: `~astropy.coordinates.BaseRepresentation` or None
```
Comment (astrofrog) [inline sunpy/coordinates/frames.py]:
```
`None`
```

(basically None, True, and False can be included in single back-ticks as their references should resolve)

File: sunpy/coordinates/frames.py [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+        A representation object or None to have no data.
+    lon: `Angle` object.
+        The longitude for this object (``lat`` must also be given and
+        ``representation`` must be None).
+    lat: `Angle` object.
+        The latitude for this object (``lon`` must also be given and
+        ``representation`` must be None).
+    rad: `astropy.units.Quantity` object.
+        This quantity holds the radial distance. If not specified, it is, by
+        default, the solar radius. Optional, must be keyword
```
Comment (astrofrog) [inline sunpy/coordinates/frames.py]: the solar radius, as defined by who? Maybe explicitly state the value and reference for that value?

File: sunpy/coordinates/frames.py [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+        A representation object. If specified, other parameters must
+        be in keyword form and if x, y and z are specified, it must
+        be None.
+    x: `Quantity` object.
+        X-axis coordinate, optional, must be keyword.
+    y: `Quantity` object.
+        Y-axis coordinate, optional, must be keyword.
+    z: `Quantity` object. Shared by both representations.
+        Z-axis coordinate, optional, must be keyword.
+    D0: `Quantity` object.
```
Comment (astrofrog) [inline sunpy/coordinates/frames.py]: In some places, I saw a mention of ``dsun`` - should make sure this is consistent

File: sunpy/coordinates/wcs_utils.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+__all__ = ['solar_wcs_frame_mapping']
+
+def solar_wcs_frame_mapping(wcs):
+    """
+    This function registers the coordinates frames to their FITS-WCS coordinate
+    type values in the `astropy.wcs.utils.wcs_to_celestial_frame` registry.
+    """
+
+    xcoord = wcs.wcs.ctype[0][0:4]
+    ycoord = wcs.wcs.ctype[1][0:4]
```
Comment (astrofrog) [inline sunpy/coordinates/wcs_utils.py]: Do you want to support cases where x and y are swapped? In Astropy, we're forcing the axes to be re-ordered for this function with:

```
    wcs = wcs.sub([WCSSUB_CELESTIAL])
```

but since I'm not sure if you can do this here, you may want to hard code trying both orders?

What if you have a spectral cube? For Astropy, this still works, again because of the above.

**Comment by astrofrog** [comment]
This looks great! I can't comment on the accuracy of the transforms obviously, but otherwise it looks nice :) I really would recommend investing a few hours coming up with some independent numerical tests, for accuracy - it'll be really worth it.

**Comment by wafels** [comment]
Turns out @MSKirk has been using a lot of co-ordinate transformations in his work.  We're collating some examples to compare sunpy results to results from SSWIDL.

**Comment by Cadair** [comment]
@wafels some comparison to hard facts would be nice ;)

**Comment by wafels** [comment]
Shouldn't the D0 property update when a dateobs is not None?

For example..

In [14]: c = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame='helioprojective')

In [15]: c
Out[15]: 
<SkyCoord (Helioprojective: D0=149597870.7 km, dateobs=None, L0=0.0 deg, B0=0.0 deg, RSun=695508.0 km): (Tx, Ty) in arcsec
    (-100.0, 500.0)>

In [17]: c = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame='helioprojective', dateobs="2015-06-18")

In [18]: c
Out[18]: 
<SkyCoord (Helioprojective: D0=149597870.7 km, dateobs=2015-06-18 00:00:00, L0=0.0 deg, B0=0.0 deg, RSun=695508.0 km):00:00, L0=0.0 deg, B0=0.0 deg, RSun=695508.0 km): (Tx, Ty) in arcsec
    (-100.0, 500.0)>

**Comment by Cadair** [comment]
@wafels Well that's making the implicit assumption the observer is at Earth.

Currently, this package makes no such assumption, and our ephemerides are not really good enough to do that anyway...

I have generally assumed that the primary use case is people creating coordinates for images where this info is in the header.

**Comment by wafels** [comment]
What are the assumptions made about the position of the observer?

So if I wanted to co-ordinates as if from Earth, I'd have to change the value of D0 myself?

**Comment by Cadair** [comment]
No assumptions are made about the position of the observer. You have to explicity specify it.

There are ways to short-cut it if you are creating coordinates from a Map, but I haven't made any modifications to Map in this PR to make that easy.

**Comment by wafels** [comment]
How do I explicitly specify the position of the observer?

**Comment by Cadair** [comment]
You have to specify B0, L0 and D0.

For example:
```python
c = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame='helioprojective', L0=1*u.deg, B0=5*u.deg, D0=1*u.au)
```

**Comment by Cadair** [comment]
@wafels I realise this is a feature we should have, I just think it is outside of the scope of this PR. We need a way of specifiying an observer to the frame, like Earth or L1 and then letting that observer "thing" take care of converting it to the parameters the SkyCoord expects.

**Comment by MSKirk** [comment]
I second that thought. An easy way to set the observer location based upon the input header is a necessity to be able to work with multiple viewpoints simultaneously.

**Comment by Cadair** [comment]
@MSKirk As I say, using Map makes it a little easier, as for multiple view points see this example: https://gist.github.com/Cadair/cc23d737f1124470e57d

**Comment by Cadair** [comment]
@MSKirk @wafels shall we talk about this at the dev meeting?

**Comment by MSKirk** [comment]
@Cadair sounds good to me.

**Comment by wafels** [comment]
absolutely.  very important when considering STEREO, Solar Probe and Solar
Orbiter.

File: sunpy/coordinates/frameattributes.py [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+            except Exception as err:
+                raise ValueError('Invalid time input {0}={1!r}\n{2}'
+                                 .format(self.name, value, err))
+            converted = True
+
+        if not out.isscalar:
+            raise ValueError('Time input {0}={1!r} must be a single (scalar) value'
+                             .format(self.name, value))
+
+        return out, converted
```
Comment (wafels) [inline sunpy/coordinates/frameattributes.py]: newline

File: doc/source/code_ref/coordinates.rst (hunk 8 above)
Comment (Cadair) [inline doc/source/code_ref/coordinates.rst]: done

File: sunpy/coordinates/wcs_utils.py (hunk 24 above)
Comment (Cadair) [inline sunpy/coordinates/wcs_utils.py]: How does `WCSSUB_CELESTIAL` work? Would it be compatible with the solar CTYPES?

File: sunpy/coordinates/frames.py (hunk 23 above)
Comment (Cadair) [inline sunpy/coordinates/frames.py]: I think all the API uses `D0`?

**Comment by ehsteve** [comment]
@Cadair just tried out your example above which is impressive but it is incredibly slow to pan or move the map around, unusably so, in my opinion. What is going on?

**Comment by ehsteve** [comment]
Also it looks like all of the original functions are still around and still use the original code such as `draw_grid`, `draw_limb`. They should be updated to use this new functionality as there should not be two ways of doing the same thing.

**Comment by ehsteve** [comment]
I would suggest that the API that is currently defined to do a transformation is rather unintuitive. From your worksheet,
```python 

hgs = hpc_aia.transform_to('heliographicstonyhurst')
hgs.D0 = maps['SECCHI'].dsun
hgs.L0 = maps['SECCHI'].heliographic_longitude
hgs.B0 = maps['SECCHI'].heliographic_latitude
hpc_B = hgs.transform_to('helioprojective')
```
The fact that the new observer location is set by updating `L0, B0, D0` in heliographic coordinates where these coordinate don't mean anything is strange and extremely unintuitive. I would suggest that this API be changed. The way the API is defined in IDL (by B. Thompson) is that you define a WCS object for where your new observer is and then transfer to that by passing it to a general transformation function.

**Comment by Cadair** [comment]
@ehsteve So taking your comments one at a time:

> @Cadair just tried out your example above which is impressive but it is incredibly slow to pan or move the map around, unusably so, in my opinion. What is going on?

This is a limitation of wcsaxes and astropy coordinates and the fact that no performance optimsation has been done on these transformations. Improving this will be on the list post-merge.

> Also it looks like all of the original functions are still around and still use the original code such as draw_grid, draw_limb. They should be updated to use this new functionality as there should not be two ways of doing the same thing.

This is deliberate, I feel as though the old functionality should be maintained as the primary for a release to make sure there are no regressions, and to give us time to work on the performance and usability of this code. Also those changes should be kept to a different PR to make this one easier to review.

> The fact that the new observer location is set by updating L0, B0, D0 in heliographic coordinates where these coordinate don't mean anything is strange and extremely unintuitive.

I see your point, but it is precisely the fact that they don't mean anything is what allows you to change them, the `SkyCoord` instance is immutable so changing those parameters while they were describing the origin of the system would violate this immutability.

> I would suggest that this API be changed. 

This API is not final at all, and I would have no problem wrapping it in the future. It is at least now possible, which is an improvement from before!!

**Comment by ehsteve** [comment]
@Cadair sound good. :+1:

File: doc/source/code_ref/coordinates.rst (hunk 2 above)
Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: I agree here. I was thinking that ``helioprojectiveradial -> helioprojective_radial```

File: doc/source/code_ref/coordinates.rst [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+
+SunPy implements support for the following solar physics coordinate systems:
+
+* Helioprojective (Cartesian) `~sunpy.coordinates.frames.HelioProjective`
+* Helioprojective (Radial) `~sunpy.coordinates.frames.HelioProjectiveRadial`
+* Heliocentric `~sunpy.coordinates.frames.HelioCentric`
+* Heliographic Stonyhurst `~sunpy.coordinates.frames.HelioGraphicStonyhurst`
+* Heliographic Carrington `~sunpy.coordinates.frames.HelioGraphicCarrington`
+
+for a complete description of these frames, see `sunpy.coordinates.frames`, for
```
Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: Capital F and use some conjunction instead of the comma.

File: doc/source/code_ref/coordinates.rst [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+a more detailed description of the frames see [(Thompson 2006)](http://dx.doi.org/10.1051/0004-6361:20054262).
```
Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: Add a link to the coordinates gist.

Comment (dpshelio) [inline doc/source/code_ref/coordinates.rst]: I would prefer to something more permanent than just a gist that can be deleted... maybe upload it to figshare or zenodo? and get the doi from there?

Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: Yeah thats a better idea.

Comment (Cadair) [inline doc/source/code_ref/coordinates.rst]: I will add the contents of the gist to the sunpy example gallery at a later date.

File: doc/source/code_ref/coordinates.rst [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+  <Quantity -72241.0 km>
+  >>> c.y
+  <Quantity 361206.1 km>
+  >>> c.z
+  <Quantity 589951.4 km>
+
+``HeliographicStonyhurst`` and ``HeliographicCarrington``
+#########################################################
+
+Both the heliographic frames use latitude, longitude and radius so are accessed as follows:
```
Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: radius which are

File: doc/source/code_ref/coordinates.rst [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+
+
+Design of the Coordinates Module
+--------------------------------
+
+This module works by defining a collection of ``Frames``
+(`sunpy.coordinates.frames`), which exists on a transformation graph, where the
+transformations between the coordinate frames are then defined and registered
+with the transformation graph (`sunpy.coordinates.transformations`). Currently,
+the SunPy frames are not transformable to the frames in Astropy, as their is no
```
Comment (nabobalis) [inline doc/source/code_ref/coordinates.rst]: there

File: sunpy/coordinates/frames.py [hunk 30]
```diff
@@ -0,0 +1,375 @@
+"""
+Common solar physics coordinate systems.
+
+This submodule implements various solar physics coordinate frames for use with
+the `astropy.coordinates` module.
+"""
+from __future__ import absolute_import, division
+
+# NumPy import
```
Comment (nabobalis) [inline sunpy/coordinates/frames.py]: Delete

File: sunpy/coordinates/frames.py [hunk 31]
```diff
@@ -0,0 +1,375 @@
...
+import numpy as np
+
+# Astropy imports
```
Comment (nabobalis) [inline sunpy/coordinates/frames.py]: Delete

File: sunpy/coordinates/frames.py [hunk 32]
```diff
@@ -0,0 +1,375 @@
...
+from astropy import units as u
+from astropy.coordinates.representation import (CartesianRepresentation,
+                                                UnitSphericalRepresentation,
+                                                SphericalRepresentation)
+from astropy.coordinates.baseframe import (BaseCoordinateFrame,
+                                           RepresentationMapping)
+from astropy.coordinates import FrameAttribute
+
+# SunPy imports
```
Comment (nabobalis) [inline sunpy/coordinates/frames.py]: Delete

File: sunpy/coordinates/representation.py [hunk 33]
```diff
@@ -0,0 +1,116 @@
+"""
+SunPy specific representations.
+
+This submodule extends `astropy.coordinates.representations` with two Spherical
+representation classes, primarily for use with the Helioprojective coordinate
+system, due to the convention of Longitude going from -180 to 180 degrees.
+"""
+
+from __future__ import absolute_import, division
+# Astropy
```
Comment (nabobalis) [inline sunpy/coordinates/representation.py]: Delete

File: sunpy/coordinates/tests/test_frames.py [hunk 34]
```diff
@@ -0,0 +1,346 @@
...
+#    # Check the attrs are correct
+#    assert hpc1.rho == 100 * u.km
+#    assert hpc1.psi == 25 * u.deg
+#    assert hpc1.distance == 1 * u.Mm
+#
+#    # Check the attrs are in the correct default units
+#    assert hpc1.rho.unit is u.km
+#    assert hpc1.psi.unit is u.arcsec
+#    assert hpc1.distance.unit is u.km
+
```
Comment (nabobalis) [inline sunpy/coordinates/tests/test_frames.py]: Why is this all commented out? Should it not be deleted?

File: sunpy/coordinates/transformations.py [hunk 35]
```diff
@@ -0,0 +1,273 @@
...
+import numpy as np
+
+# Astropy imports
+from astropy import units as u
+from astropy.coordinates.representation import (CartesianRepresentation,
+                                                UnitSphericalRepresentation)
+from astropy.coordinates.baseframe import frame_transform_graph
+from astropy.coordinates.transformations import FunctionTransform
+
+# SunPy imports
```
Comment (nabobalis) [inline sunpy/coordinates/transformations.py]: Delete

File: sunpy/coordinates/transformations.py [hunk 36]
```diff
@@ -0,0 +1,273 @@
...
+
+This module contains the functions for converting one
+`sunpy.coordinates.frames` object to another.
+"""
+from __future__ import absolute_import, division
+
+# NumPy import
+import numpy as np
+
+# Astropy imports
```
Comment (nabobalis) [inline sunpy/coordinates/transformations.py]: Delete

File: sunpy/coordinates/transformations.py [hunk 37]
```diff
@@ -0,0 +1,273 @@
+# -*- coding: utf-8 -*-
+"""
+Coordinate Transformation Functions
+
+This module contains the functions for converting one
+`sunpy.coordinates.frames` object to another.
+"""
+from __future__ import absolute_import, division
+
+# NumPy import
```
Comment (nabobalis) [inline sunpy/coordinates/transformations.py]: Delete

File: sunpy/coordinates/transformations.py [hunk 38]
```diff
@@ -0,0 +1,273 @@
...
+    Generates a string for use with the coordinate package's docstring
+    to show the available transforms and coordinate systems
+    """
+    import inspect
+    from textwrap import dedent
+    from sunpy.extern import six
+    from astropy.coordinates.baseframe import (BaseCoordinateFrame,
+                                               frame_transform_graph)
+
+    # TODO: Make this just show the SunPy Frames
```
Comment (nabobalis) [inline sunpy/coordinates/transformations.py]: Do it now?

File: sunpy/coordinates/wcs_utils.py (hunk 24 above)
Comment (nabobalis) [inline sunpy/coordinates/wcs_utils.py]: Any update here?

**Comment by nabobalis** [comment]
Looks fine to me bar the little items. 

Was there any decision about on an easier way to change the observer, or will that come in a future PR after this gets more integration with Map?

**Comment by Cadair** [comment]
## Todo:

* [ ] Make sure HPR works or remove it from the API.
* [ ] Change the string names to have underscores in.
* [ ] Look into `WCSSUB_CELESTIAL`

File: sunpy/coordinates/wcs_utils.py (hunk 24 above)
Comment (astrofrog) [inline sunpy/coordinates/wcs_utils.py]: Looking at the WCSLIB source code (https://github.com/astropy/astropy/blob/master/cextern/wcslib/C/wcs.c#L2438) I think that ``WCSSUB_CELESTIAL`` is defined as an alias for ``WCSSUB_LONGITUDE`` and ``WCSSUB_LATITUDE``, and these are defined as anything that matches:

* ``RA--*``, ``?LON*``, or ``??LN*`` for the longitude
* ``DEC-*``, ``?LAT*``, or ``??LT*`` for the latitude

Does this covert the solar ctypes adequately? (it won't match e.g. heliocentric cartesian, but that's normal)

Comment (Cadair) [inline sunpy/coordinates/wcs_utils.py]: it dosen't cover every case in this function, but it might be best to try that first, as it will handle the swapping.

File: sunpy/coordinates/tests/test_frames.py (hunk 34 above)
Comment (Cadair) [inline sunpy/coordinates/tests/test_frames.py]: it needs to be changed to test the rather edge case of heliocentric radial.

**Comment by nabobalis** [comment]
Thumbs up from me.

**Comment by ayshih** [comment]
Changelog entry!

:+1:

**Comment by Cadair** [comment]
@ayshih added

**Comment by Cadair** [comment]
SOMEBODY PRESS THE BUTTON

**Comment by ehsteve** [comment]
Done! Please blame @ayshih for my pressing the button.

**Comment by Cadair** [comment]
PARTY PARTY

**Comment by khughitt** [comment]
Nice work everyone! :beers:

**Comment by Cadair** [comment]
Hi! @khughitt We merged it at last.

**Comment by eteq** [comment]
You know, I still plan to review this someday.  But I guess I'll have to actually do it in master now ;)

**Comment by Cadair** [comment]
It works, it has been released, and it allows you to do fun things like this: http://docs.sunpy.org/en/stable/generated/gallery/tutorials/SDO_to_STEREO_Coordinate_Conversion.html#sphx-glr-generated-gallery-tutorials-sdo-to-stereo-coordinate-conversion-py

Feedback still always welcome :)
